source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,312 @@
+#pragma once
+
+#include <queue>
+
+#include "envoy/api/v2/discovery.pb.h"
+#include "envoy/common/token_bucket.h"
+#include "envoy/config/grpc_mux.h" // for ControlPlaneStats
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/backoff_strategy.h"
+#include "common/common/logger.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * TODO SOMETHING SOMETHING subscription. Also handles per-xDS API stats/logging.
+ */
+template <class ResourceType>
+class IncrementalSubscriptionImpl
+    : public IncrementalSubscription<ResourceType>,
+      Grpc::TypedAsyncStreamCallbacks<envoy::api::v2::IncrementalDiscoveryResponse>,
+      Logger::Loggable<Logger::Id::config> {
+public:
+  IncrementalSubscriptionImpl(const LocalInfo::LocalInfo& local_info,
+                              Grpc::AsyncClientPtr async_client, Event::Dispatcher& dispatcher,
+                              const Protobuf::MethodDescriptor& service_method,
+                              Runtime::RandomGenerator& random,
+                              const RateLimitSettings& rate_limit_settings,
+                              IncrementalSubscriptionStats stats,
+                              ControlPlaneStats control_plane_stats)
+      : type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        local_info_(local_info), async_client_(std::move(async_client)),
+        service_method_(service_method), rate_limiting_enabled_(rate_limit_settings.enabled_),
+        stats_(stats), control_plane_stats_(control_plane_stats) {
+    retry_timer_ = dispatcher.createTimer([this]() -> void { establishNewStream(); });
+    if (rate_limiting_enabled_) {
+      // Default Bucket contains 100 tokens maximum and refills at 10 tokens/sec.
+      limit_request_ = std::make_unique<TokenBucketImpl>(
+          rate_limit_settings.max_tokens_, dispatcher.timeSystem(), rate_limit_settings.fill_rate_);
+      drain_request_timer_ = dispatcher.createTimer([this]() -> void { drainRequests(); });
+    }
+    backoff_strategy_ =
```

Comment:
I think there is a lot of boiler plate similarity between this and `SubscriptionImpl`, e.g. the rate limiting, tracking connection status in stats, etc.. Since these are templates, it's not practical to do inheritance, but maybe we could create a shared member object that handles common subscription concerns like creating a gRPC stream and rate limiting it? I'm uncomfortable with potential drift between these two implementations.

Reply:
I understand, that makes sense; will do.

Reply:
(#5681)